File: python/validate_bst/validate_bst.py
from typing import Optional

from treenode.treenode import TreeNode


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def dfs(node, min_value, max_value):
            if not node:
                return True

            if not (min_value < node.val < max_value):
                return False

            return dfs(node.left, min_value, node.val) and dfs(node.right, node.val, max_value)

        return dfs(root, -float("inf"), float("inf"))

File: go/validateBST/validate_bst.go
package main

import "math"

type TreeNode struct {
	Val   int32
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	return dfs(root, math.MinInt32, math.MaxInt32)
}

func dfs(node *TreeNode, minValue int32, maxValue int32) bool {
	if node == nil {
		return true
	}

	if !(node.Val > minValue && node.Val < maxValue) {
		return false
	}

	return dfs(node.Left, minValue, maxValue) && dfs(node.Right, node.Val, maxValue)
}
